File: scripts/validate_data.py
#!/usr/bin/env python3
import json
import sys
from pathlib import Path

REQUIRED_TOP_LEVEL = {"client", "period", "freshness", "totals", "daily", "insights"}
REQUIRED_TOTALS = {"spend", "impressions", "clicks", "conversions", "revenue", "sessions"}
REQUIRED_DAILY = {"date", "source", "campaign", "spend", "impressions", "clicks", "conversions", "revenue"}


def fail(message: str) -> None:
    print(f"DATA_VALIDATION_ERROR: {message}", file=sys.stderr)
    raise SystemExit(1)


def validate(path: Path) -> None:
    data = json.loads(path.read_text())
    missing = REQUIRED_TOP_LEVEL - set(data)
    if missing:
        fail(f"missing top-level keys: {sorted(missing)}")
    missing_totals = REQUIRED_TOTALS - set(data["totals"])
    if missing_totals:
        fail(f"missing totals keys: {sorted(missing_totals)}")
    if not data["daily"]:
        fail("daily rows cannot be empty")
    for i, row in enumerate(data["daily"]):
        missing_daily = REQUIRED_DAILY - set(row)
        if missing_daily:
            fail(f"daily row {i} missing keys: {sorted(missing_daily)}")
        for numeric in ["spend", "impressions", "clicks", "conversions", "revenue"]:
            if not isinstance(row[numeric], (int, float)):
                fail(f"daily row {i} has non-numeric {numeric}")
    text = path.read_text().lower()
    forbidden = ["windsor_api_key", "api_key=", "netlify_auth_token", "github_token", "openai_api_key"]
    leaked = [needle for needle in forbidden if needle in text]
    if leaked:
        fail(f"possible secret marker found: {leaked}")
    print(f"OK: {path} contains {len(data['daily'])} daily rows and passed schema/secret checks")


if __name__ == "__main__":
    validate(Path(sys.argv[1] if len(sys.argv) > 1 else "data/demo/latest.json"))

File: scripts/validate-data.mjs
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

const REQUIRED_TOP_LEVEL = ['client', 'period', 'freshness', 'totals', 'daily', 'insights'];
const REQUIRED_TOTALS = ['spend', 'impressions', 'clicks', 'conversions', 'revenue', 'sessions'];
const REQUIRED_DAILY = ['date', 'source', 'campaign', 'spend', 'impressions', 'clicks', 'conversions', 'revenue'];
const FORBIDDEN_SECRETS = ['windsor_api_key', 'api_key=', 'netlify_auth_token', 'github_token', 'openai_api_key'];

function fail(message) {
  console.error(`DATA_VALIDATION_ERROR: ${message}`);
  process.exit(1);
}

function validate(filePath) {
  const absolutePath = resolve(filePath);
  let content;
  try {
    content = readFileSync(absolutePath, 'utf-8');
  } catch (err) {
    fail(`Could not read file ${absolutePath}: ${err.message}`);
  }

  let data;
  try {
    data = JSON.parse(content);
  } catch (err) {
    fail(`Invalid JSON in ${absolutePath}: ${err.message}`);
  }

  for (const key of REQUIRED_TOP_LEVEL) {
    if (!(key in data)) {
      fail(`missing top-level key: ${key}`);
    }
  }

  for (const key of REQUIRED_TOTALS) {
    if (!(key in data.totals)) {
      fail(`missing totals key: ${key}`);
    }
  }

  if (!Array.isArray(data.daily) || data.daily.length === 0) {
    fail('daily rows cannot be empty');
  }

  data.daily.forEach((row, index) => {
    for (const key of REQUIRED_DAILY) {
      if (!(key in row)) {
        fail(`daily row ${index} missing key: ${key}`);
      }
    }
    for (const numeric of ['spend', 'impressions', 'clicks', 'conversions', 'revenue']) {
      if (typeof row[numeric] !== 'number' || Number.isNaN(row[numeric])) {
        fail(`daily row ${index} has non-numeric ${numeric}`);
      }
    }
  });

  const lowerContent = content.toLowerCase();
  for (const needle of FORBIDDEN_SECRETS) {
    if (lowerContent.includes(needle)) {
      fail(`possible secret marker found: ${needle}`);
    }
  }

  console.log(`OK: ${filePath} contains ${data.daily.length} daily rows and passed schema/secret checks`);
}

const targetFile = process.argv[2] || 'data/demo/latest.json';
validate(targetFile);
